Default concourse pipeline

siredmar opened the pull request.

Commit:
#345: First pipeline test

Commit:
#345: first working version of versioning in pipeline

Commit:
#345: Added convenience for using inputs and outputs for gitversion

Commit:
#345: Added local config, pipeline file configurable with config

Commit:
#345: Added readme and example

Commit:
#345: Updated title to integration guide readme

Commit:
#345: Modified Readme regarding credentials, modified s3 local credentials

mlilien commented on README.md:
```diff
@@ -21,6 +21,20 @@ The separation between meta.yaml and dobi.yaml is necessary to integrate the bui
+
+The pipeline file must be uploaded to concourse CI via `fly`. 
+Enter the build users ssh private key into the file `ci/credentials.template.yaml` and rename it to `ci/credentials.yaml`. 
+**Note, that files calld `credentials.yaml` will is ignored by `.gitignore` and will not be checked in.**
```
typos + grammar issues.

siredmar replied:
Sorry for that. A hamster jumped on my keyboard.

mlilien commented on README.md:
```diff
@@ -21,6 +21,20 @@ The separation between meta.yaml and dobi.yaml is necessary to integrate the bui
+
+After sucessfully uploading the pipeline to concourse CI login and unpause it. After that the pipeline should be triggered by new commits on the master branch (or new tags if enabled in `pipeline.yaml`).
+
+If you want to integrate this into another building block you should have a look at the [integration documentation](ci/example/README.md).
```
"," missing
-> If you want to integrate this into another building block, you should ...

siredmar replied:
Ok

mlilien commented on README.md:
```diff
@@ -21,6 +21,20 @@ The separation between meta.yaml and dobi.yaml is necessary to integrate the bui
+
+    $ fly -t <target> set-pipeline -n -p bb-buildingblock -l ci/config.yaml -l ci/credentials.yaml -c ci/pipeline.yaml
+
+After sucessfully uploading the pipeline to concourse CI login and unpause it. After that the pipeline should be triggered by new commits on the master branch (or new tags if enabled in `pipeline.yaml`).
```
sucessfully -> successfully

siredmar replied:
hit me!

mlilien commented on README.md:
```diff
@@ -21,6 +21,20 @@ The separation between meta.yaml and dobi.yaml is necessary to integrate the bui
+The pipeline file must be uploaded to concourse CI via `fly`. 
+Enter the build users ssh private key into the file `ci/credentials.template.yaml` and rename it to `ci/credentials.yaml`. 
+**Note, that files calld `credentials.yaml` will is ignored by `.gitignore` and will not be checked in.**
+In further releases there will be an key value store to keep track of the users credentials.
```
"an key" -> "a key"

siredmar replied:
yes

mlilien commented on ci/credentials.template.yaml:
```diff
@@ -0,0 +1,9 @@
+---
+# Put here your fill private key
```
grammar

siredmar replied:
no grammar but nonsense. fixed.

mlilien commented on ci/example/README.md:
```diff
@@ -0,0 +1,96 @@
+
+# Concourse pipeline integration guide
+
+For using this concourse CI popeline it is recommended to have two config files:
```
"popeline" -> "pipeline"

siredmar replied:
white smoke rises... fixed.

mlilien commented on ci/example/README.md:
```diff
@@ -0,0 +1,96 @@
+
+For using this concourse CI popeline it is recommended to have two config files:
+
+-   `config.yaml`: for production pruposes
```
"pruposes" -> "purposes"

siredmar replied:
sounds funny, fixed.

mlilien commented on ci/example/README.md:
```diff
@@ -0,0 +1,96 @@
+
+# Example concourse CI pipeline
+
+This example is inteneded to make integration of the build pipeline more easy into other building blocks.
```
-> "This example is intended to show the integration of a build pipeline in a  derived building block."???

siredmar replied:
sounds good.

mlilien commented on ci/example/README.md:
```diff
@@ -0,0 +1,96 @@
+## Adaptions for your building block
+
+This section describes what needs to be done within the several files. 
+**Note, that files calld `credentials.yaml` will is ignored by `.gitignore` and will not be checked in.**
```
copy paste? typo + grammar

siredmar replied:
indeed. fixed.

mlilien commented on ci/example/README.md:
```diff
@@ -0,0 +1,96 @@
+### config.example.yaml
+
+Open the `config.example.yaml` and specify the fields that are described.
+This config file shows an example for an fictive building block calld `bb-example`.
```
"an fictive" -> "a fictive"
"calld" -> "called"

siredmar replied:
"do you want do make it double?"
fixed.

mlilien commented on ci/example/credentials.example.yaml:
```diff
@@ -0,0 +1,9 @@
+---
+# Put here your fill private key
```
grammar/sense

siredmar replied:
It makes no sense, but what does?
Fixed.

Commit:
#345: fixed typos and grammar

Commit:
#435: updated Readme, fixed typos

Commit:
#435: Added configurable docker image artifact output destination

Commit:
#345: automatically adding s3 bucket if not existent

FlorianReinhold commented on .gitignore:
```diff
@@ -3,4 +3,5 @@ build
-.dobi
\ No newline at end of file
+.dobi
+ci/config.local.yaml
```
duplicated. see ci/.gitignore

siredmar replied:
yes

FlorianReinhold commented on ci/config.yaml:
```diff
@@ -0,0 +1,29 @@
+s3_access_key: localaccess
+s3_secret_key: localsecret
+## This enables/disables the ssl verification for the S3 endpoint
+s3_disable_ssl: false
```
cool :+1:

FlorianReinhold commented:
i like the quality of the pr's after a first review. i would only choose the directory structure a bit different. with the current structure it would be very difficult to update the bb-buildingblock template in derived blocks. for me the pipeline file is not an example but a template that i would expect on the highest level.

siredmar commented:
> i like the quality of the pr's after a first review. i would only choose the directory structure a bit different. with the current structure it would be very difficult to update the bb-buildingblock template in derived blocks. for me the pipeline file is not an example but a template that i would expect on the highest level.

For the directory structure: i like the idea of having the ci/cd stuff concentrated in the ci folder because the pipeline gets sent to concourse via a manual step using the command `fly`. 
Derived blocks also contain this `ci` folder and have an example how to integrate the main pipeline in their bb's.

You're suggesting that i put the pipeline.yaml (and all needed files like config.yaml?) in the bb's root but the benefit is unclear to me. Derived bb's still have to adapt the pipeline.yaml to their needs, the user still has to set it up with `fly`. Could you elaborate more?

Commit:
#345: moved pipeline.yaml to bb's root folder

- put S3 credentials in credentials.yaml file
- incremented bb-gitversion to 0.2.0

FlorianReinhold reviewed (commented):
> > i like the quality of the pr's after a first review. i would only choose the directory structure a bit different. with the current structure it would be very difficult to update the bb-buildingblock template in derived blocks. for me the pipeline file is not an example but a template that i would expect on the highest level.
> 
> For the directory structure: i like the idea of having the ci/cd stuff concentrated in the ci folder because the pipeline gets sent to concourse via a manual step using the command `fly`.
> Derived blocks also contain this `ci` folder and have an example how to integrate the main pipeline in their bb's.
> 
> You're suggesting that i put the pipeline.yaml (and all needed files like config.yaml?) in the bb's root but the benefit is unclear to me. Derived bb's still have to adapt the pipeline.yaml to their needs, the user still has to set it up with `fly`. Could you elaborate more?

Basically it is about how the buildscripts are placed. If we place dobi or e.g. cmake scripts, they are put into subfolders of the components to be translated. If concourse has the possibility to include different subsripts, they will be placed in the subfolders as well.
This would result in an inconsistent structure, the "root" document would be placed in a subfolder "ci", the other scripts would be spread over the project. Therefore I suggest to put the root document in the project root and to put only generic parts in a subfolder "ci".

mlilien commented on ci/.gitignore:
```diff
@@ -0,0 +1,2 @@
+credentials.yaml
```
i don't see this line in https://github.com/siredmar/bb-buildingblock/blob/ci/.gitignore and locally?!

mlilien replied:
solved ... it's another file

mlilien commented on README.md:
```diff
@@ -19,7 +19,22 @@ By default three dobi resources are predefined (but not implemented):
+
+The pipeline file must be uploaded to concourse CI via `fly`. 
+Enter the build users ssh private key into the file `ci/credentials.template.yaml` and rename it to `ci/credentials.yaml`. 
+**Note: Files called `credentials.yaml` are ignored by `.gitignore` and will not be checked in.**
```
reads funny.

-> **Note:  `credentials.yaml` is ignored by `.gitignore` and will not be checked in.**

siredmar replied:
Modified.

mlilien commented on ci/example/README.md:
```diff
@@ -0,0 +1,100 @@
+## Adaptions for your building block
+
+This section describes what needs to be done within the several files. 
+**Note: Files called `credentials.yaml` are ignored by `.gitignore` and will not be checked in.**
```
reads funny.

-> **Note:  `credentials.yaml` is ignored by `.gitignore` and will not be checked in.**

siredmar replied:
Modified.

mlilien commented on README.md:
```diff
@@ -19,7 +19,22 @@ By default three dobi resources are predefined (but not implemented):
+
+After successfully uploading the pipeline to concourse CI login and unpause it. After that the pipeline should be triggered by new commits on the master branch (or new tags if enabled in `pipeline.yaml`).
+
+If you want to integrate the CI/CD pipeline into another building block, you should have a look at the [integration documentation](ci/example/README.md).
```
ImHo it should be moved to example/ci/README.md

siredmar replied:
You are absolutely right. Moved.

mlilien commented on pipeline.yaml:
```diff
@@ -0,0 +1,85 @@
+                if [ $? != 0 ]; then aws s3 mb s3://((name)) --endpoint-url ((s3_endpoint)); fi
+      - put: s3-gitversion
+        params:
+          file: gen/gitversion.tar.gz
```
we overwrite this file on each pipeline run ... this is okay for manual triggering when there is no version change.
But what on version change ... ? Either the version information itself should be part of the bucketname or the version file.

mlilien replied:
we aggreed to use the version as part of the bucketname

siredmar replied:
S3 bucket names will now be created with SemVer suffix e.g. bb-example-1.2.3

siredmar replied:
After further consideration the s3 bucket will be created without the version suffix. Instead, the generated gitversion archive file gets either the FullSemVer suffix for `master` and a InformationalVersion as suffix. These Version strings are slightly modified to allow the usage as docker tags ('+' replaced by '-') .

Commit:
#345: overall updates, see description

- dobi.sh using gitversion env file
- s3 bucket created with SemVer suffix
- reordered example
- Integration guide adaptions to make things more clear
- several typo fixes

Commit:
#345: fixed typo, fixed s3 bucket creation

Commit:
#345: gitversion versioned in bucket

siredmar commented:
master branch shall receive FullSemVer e.g. `0.1.0+2`
development branches shall receive InformationalVersion e.g. `0.3.10+2.Branch.master.Sha.9e7005a4be1f0de16550963a1b9784981224506a`

Edit: see https://github.com/elbb/bb-buildingblock/pull/13#discussion_r429205692

Commit:
#345: master branch gets SemVer, all other branches InformationalVersion

Commit:
#345: added user/pw for docker registry, replaced '+' in InformationalVersion with '-'

mlilien commented on README.md:
```diff
@@ -4,22 +4,40 @@
+## Using concourse CI for a CI/CD build
+
+The pipeline file must be uploaded to concourse CI via `fly`. 
+Enter the build users ssh private key into the file `ci/credentials.template.yaml` and rename it to `ci/credentials.yaml`. 
```
path of credentials.template is incorrect

siredmar replied:
added the file

mlilien commented on example/ci/pipeline.yaml:
```diff
@@ -0,0 +1,118 @@
+                else 
+                  cat gen/plain/InformationalVersion > gen/version
+                fi &&
+                sed 's/+/-/g' gen/version > gen/version &&
```
doesnt work in my concourse env. the file gen/version is empty afterwards. I had to change the line to 
```
sed -i 's/+/-/g' gen/version &&
```

siredmar replied:
modified

mlilien commented on pipeline.yaml:
```diff
@@ -0,0 +1,94 @@
+                else 
+                  cat gen/plain/InformationalVersion > gen/version
+                fi &&
+                sed 's/+/-/g' gen/version > gen/version &&
```
doesnt work in my concourse env. the file gen/version is empty afterwards. I had to change the line to 
```
sed -i 's/+/-/g' gen/version &&
```

siredmar replied:
modified

mlilien commented on example/ci/pipeline.yaml:
```diff
@@ -0,0 +1,118 @@
+      bucket: ((bb_name))
+      endpoint: ((s3_endpoint))
+      # at least one regex capture group is needed here
+      regexp: version/gitversion-(.*).tar.gz
```
i get s3-gitversion check errors, because
```regexp: version/gitversion-(.*).tar.gz``` doesnt match ```file: gen/gitversion-*.tar.gz ```

(I solved that in my bb-gitversion MR, where i names the director `gitversion` and adapted `regexp` + `file`.)

siredmar replied:
good catch! Thanks. fixed.

Commit:
#345: fixed some regex bugs, added credentials.template.yaml to ci/

mlilien commented on version.yaml:
```diff
@@ -18,7 +18,7 @@ mount=mount-buildingblock-gitversion-gen-dir:
 image=image-buildingblock-gitversion:
   image: elbb/bb-gitversion
-  tags: ["0.1.0"]
+  tags: ["0.2.0"]
```
0.3.0 ?

siredmar replied:
Not part of this PR, but i can change it.

Commit:
#345: Incremented version of bb-gitversion for dobi versioning to 0.3.0
